### DISCUSSION TITLE
Potential Bug: Space group not maintained during cell optimization despite KEEP_SPACE_GROUP=T

### DISCUSSION THREAD

**Discussion by zhangbw20** [cp2k/cp2k#5637] [General]
Dear developers of cp2k, Greetings
### Description
I encountered an issue where the space group was not maintained during lattice optimization despite setting "KEEP_SPACE_GROUP" to true. This issue was observed while optimizing the lattice structures of iodine (No.64,Cmce) and silicon (No.227,Fd-3m) in their primitive cell forms, where the non-zero elements of three upper-triangular elements in lattice were left unchanged during the whole process of optimization.
```
$ grep -A2 "Vector a" I2_N01.out
 CELL| Vector a [angstrom]:       3.763    -2.281     0.000   |a| =     4.400496
 CELL| Vector b [angstrom]:       3.763     2.281     0.000   |b| =     4.400496
 CELL| Vector c [angstrom]:       0.000     0.000     9.706   |c| =     9.706146
--
 CELL| Vector a [angstrom]:       3.802    -2.281     0.000   |a| =     4.433490
 CELL| Vector b [angstrom]:       3.802     2.325     0.000   |b| =     4.456413
 CELL| Vector c [angstrom]:       0.000     0.000     9.693   |c| =     9.692778
--
 CELL| Vector a [angstrom]:       4.028    -2.281     0.000   |a| =     4.629319
 CELL| Vector b [angstrom]:       4.028     2.575     0.000   |b| =     4.781154
 CELL| Vector c [angstrom]:       0.000    -0.000     9.604   |c| =     9.603716
--
 CELL| Vector a [angstrom]:       4.269    -2.281     0.000   |a| =     4.840494
 CELL| Vector b [angstrom]:       4.269     2.825     0.000   |b| =     5.119503
 CELL| Vector c [angstrom]:       0.000    -0.000     9.495   |c| =     9.494987
--
 CELL| Vector a [angstrom]:       4.487    -2.281     0.000   |a| =     5.033592
 CELL| Vector b [angstrom]:       4.487     3.034     0.000   |b| =     5.416581
 CELL| Vector c [angstrom]:       0.000    -0.000     9.384   |c| =     9.383704
--
...
```

### Relevant code
Line 129~137 of gopt_f_methods.f:
```
            ALLOCATE (x0(3*nparticle + 6))
            CALL pack_subsys_particles(subsys=subsys, r=x0)
            idg = 3*nparticle
            DO i = 1, 3
               DO j = 1, i
                  idg = idg + 1
                  x0(idg) = cell%hmat(j, i)
               END DO
            END DO
```
The three upper-triangular elements were not read into x0 during the initialization of x0 which is later updated by dr for geoopt_bfgs function in bfgs_optimizer.F. 
I didn't go through all the later functions to reassure whether there were other symmetry operations implemented for x0, but the output file results would show that the three upper-triangular elements were left unchanged so that the space group can't be maintained if any of the three upper-triangular elements are non-zero (which are common for primitive cell of C-centered orthorhombic lattice, FCC, etc.). 

### Potential Impacts
Thus, maybe cp2k in its current version (cp2k-2024.1) is malfunctioned for some primitive cell optimization. Additionally, users (like me) using primitive cell for band structure calculation might suffer from the cell relaxation, especially recently RI-HFX techniques were employed which are especially suitable for small crystal's band calculations with a great number of K points. (Certainly, this could be evaded by using a conventional cell for cell optimization and then use primitive cell for other properties calculation.)

### Steps to Reproduce
Two test cases from ./QS/regtest-spgr/quartz.inp with minimal modifications are attached as postscripts. 
[I2_N01.inp.txt](https://github.com/cp2k/cp2k/files/15177573/I2_N01.inp.txt)
[Si_N01.inp.txt](https://github.com/cp2k/cp2k/files/15177576/Si_N01.inp.txt)
version: cp2k-2024.1, spglib-v1.16.2 (latest as today)

### Questions
Are there any known workarounds for this issue?

### Notes
1. Note that the identify_space_group function (calling spgr_create) in space_groups.F uses the full hmat (all 3-by-3 elements) so that it could correctly identify the space group of a conventional unit cell (or primitive cell) by spglib (v1.16.2) in both example cases. 
Line 228 of space_groups.F:
`     spgr%space_group_number = spg_get_international(spgr%international_symbol, TRANSPOSE(cell%hmat), tmp_coor, tmp_types, &
                                                      spgr%n_atom_sym, eps_symmetry)`

2. By the way, are there any other better ways than modifying the src code to WRITE outputs and recompiling for cp2k debug? 
I found that: 
a) the stacks of function being called are not fully written out into the output file so it would be very hard to see the layer structure and order of all the executing functions even if the TRACE tag in input file was set to T; 
b) I followed "debug build" instructions in https://manual.cp2k.org/trunk/CP2K_INPUT/DEBUG.html, but still couldn't let gdb TUI find source code; 
c) So using WRITE sentences and recompiling is my final choice to read the values of x0 to finally locate the problem, which took me quite a lot of time. 

3. Earlier discussions are at https://groups.google.com/g/cp2k/c/Xwna6hLwWzk  Thanks a lot for Prof. Hutter's discussions and instructions for me.

**Comment by he-zilong** [comment]
The documentations for both `&FORCE_EVAL/&SUBSYS/&CELL` and `&MOTION/&CELL_OPT` have been adjusted in #4974 and #5086 respectively with emphasis on the cell vector matrix being a lower triangular matrix. It looks less like a bug but more like a deep-rooted (see e.g. commit 14dc25a for early development) program convention that was not conveyed very well, perhaps having something to do with available degrees of freedom for cell vectors in the optimization algorithm (which hopefully some other developers can explain better). Note that the native, `spglib`-independent keywords `KEEP_ANGLES` and `KEEP_SYMMETRY` in `SUBROUTINE apply_cell_constraints()` from `src/motion/cell_opt_utils.F` currently operate under the same assumption. It may take highly non-trivial efforts to revamp the source code and rectify the behavior.

> ... if any of the three upper-triangular elements are non-zero (which are common for primitive cell of C-centered orthorhombic lattice, FCC, etc.).

It is certainly always possible to apply linear transformations in the modelling stage to specify an equivalent cell where the first cell vector A lies along the X-axis and the second cell vector B is on the XY-plane, with which the atomic coordinates and external pressure tensors are defined.

**Comment by mkrack** [comment]
The original focus of CP2K was on MD simulations, where molecular or space group symmetries have no or little relevance. While tools like `molsym` and `spglib` were added later, they currently only detect symmetries rather than enforcing them as constraints with the exception of `KEEP_SPACE_GROUP`. Recent functionalities like k-points and band structure calculations were not envisioned during the early development phase.

The convention of aligning lattice vector `a` with the x-axis and placing `b` in the xy-plane is necessary for the non-redundant stress tensor (just 6 elements) needed in direct cell optimisations (CELL_OPT). Additionally, IIRC, this convention became essential with the introduction of the DCD format to ensure a unique mapping between real and scaled atomic coordinates. In this respect, The base design of CP2K is different to conventional PW codes for solid state calculations and as a result, a one-to-one mapping of concepts between the two can be cumbersome.

**Comment by he-zilong** [comment]
The latest CP2K version will emit warnings about unconventional cell vectors, first when reading input
https://github.com/cp2k/cp2k/blob/e887d2c5ab82154dc0146e0cbeaa9f188aefe88b/src/cell_methods.F#L457-L461
and second when applying cell symmetry
https://github.com/cp2k/cp2k/blob/e887d2c5ab82154dc0146e0cbeaa9f188aefe88b/src/cell_methods.F#L508-L515
These messages could have been useful for a case like #4170 .

On a closer look, the Si_N01.inp.txt given in the first post contains fractional coordinates as indicated by keyword `SCALED T`.
```
     &CELL
            A     3.865739    3.865739   0.000000
            B     0.000000    3.865739   3.865739
            C     3.865739    0.000000   3.865739
       PERIODIC  XYZ
       MULTIPLE_UNIT_CELL  1 1 1
     &END CELL
     &COORD
            Si    0.00000    0.00000    0.00000
            Si    0.25000    0.25000    0.25000
        SCALED T
     &END COORD
```
Instead of vectors `A`, `B`, `C`, it is recommended to give the lengths `ABC` and angles `ALPHA_BETA_GAMMA` (here 5.466981 = 3.865739 * sqrt(2) for the primitive cell) so that CP2K can construct the cell vectors conforming to its program convention based on which atomic coordinates are defined. Also, `&MOTION/&CELL_OPT/KEEP_SYMMETRY T` will not be effective without defining `&FORCE_EVAL/&SUBSYS/&CELL/SYMMETRY` explicitly to something else that is not the default triclinic, because both `SYMMETRY` and `KEEP_SYMMETRY` work by asjusting the matrix of cell vectors and gradients under the program convention.
```
     &CELL
            ABC 5.466981 5.466981 5.466981
            ALPHA_BETA_GAMMA 60.0 60.0 60.0
            SYMMETRY RHOMBOHEDRAL
       PERIODIC  XYZ
       MULTIPLE_UNIT_CELL  1 1 1
     &END CELL
     &COORD
            Si    0.00000    0.00000    0.00000
            Si    0.25000    0.25000    0.25000
        SCALED T
     &END COORD
```

**Comment by he-zilong** [comment]
As the explanations above and documentations added since then have clarified the point about the program convention for cells, I am now archiving this issue and converting it to a discussion.

There are further discussions at #5310 about how a non-canonical cell should be handled.

**Comment by DCM-Uni-Paderborn** [comment]
A task-specific safety fix is proposed in #5648.

`CELL_OPT` optimizes six independent cell-matrix components and therefore assumes the canonical orientation where A lies on the X axis and B lies in the XY plane. The PR adds an early check at `CELL_OPT` startup:

- a noncanonical cell with the default `CELL%CANONICALIZE AUTO` stops with guidance to use `CANONICALIZE TRUE` or `ABC` plus `ALPHA_BETA_GAMMA`;
- explicit `CANONICALIZE FALSE` retains the legacy unconstrained behavior with a coordinate-frame-dependence warning;
- noncanonical cells combined with cell constraints, `KEEP_ANGLES`, `KEEP_SPACE_GROUP`, `KEEP_SYMMETRY`, or `KEEP_VOLUME` are rejected even with `CANONICALIZE FALSE`.

The original primitive silicon setup is included as a regression using `CANONICALIZE TRUE` and `KEEP_SPACE_GROUP T`; it completes `CELL_OPT` while preserving space group 227. The CP2K precommit, misc, and pdbg regtest checks are green.

For current input files, the recommended workaround is therefore to set `FORCE_EVAL%SUBSYS%CELL%CANONICALIZE TRUE` for general A/B/C vectors, or to specify the equivalent canonical cell through lengths and angles.

**Comment by Growl1234** [comment]
Would it be possible to separate the fixed overall orientation from the six non-redundant cell-deformation degrees of freedom, and optimize the latter in a canonical frame? In other words, the optimizer could internally work with a canonical six-parameter representation while preserving the overall orientation of the input cell when reconstructing the actual cell matrix. This would avoid introducing three redundant rotational degrees of freedom without requiring the user’s cell to be physically reoriented.

And again, I'm still not clear about semantics of `CANONICALIZE AUTO`...

**Comment by DCM-Uni-Paderborn** [comment]
Yes, this is mathematically possible, and I agree that it would be the cleaner
long-term model.

Given an input cell \(H_0\) and its equivalent canonical representation \(C_0\),
one can determine a fixed orthogonal transformation \(R\). The optimizer could
then work exclusively with the six independent components of the canonical cell
\(C\), while the cell in the original user frame would be reconstructed as

\[
H_{\mathrm{user}} = R C .
\]

Coordinates, forces, stresses, external pressure tensors, and other directional
quantities would have to be transformed consistently using the same fixed
rotation.

I think an important implementation detail is that CP2K should ideally keep the
cell canonical internally and reconstruct the original orientation only at
user-facing input/output boundaries. Passing the reconstructed noncanonical
matrix back into the force evaluation would retain the downstream assumptions
that motivated canonicalization in the first place.

Some of the required infrastructure already exists from #5307:
CP2K stores the input-to-canonical transformation and applies it to several
cell-dependent inputs. Extending this model to `CELL_OPT` would additionally
require consistent handling in optimizer packing/unpacking, cell gradients and
stress, constraints, space-group operations, trajectories, and restarts.
Therefore, I see #5648 as a safety fix rather than the final architectural
solution. The orientation-preserving six-parameter representation could be a
separate follow-up.

Regarding `CANONICALIZE AUTO`, the current implementation is simpler than its
documentation may suggest:

- cells specified through `ABC` and `ALPHA_BETA_GAMMA` are canonical by
  construction;
- already canonical `A/B/C` matrices remain unchanged;
- a general noncanonical matrix is kept in its input orientation and produces a
  warning;
- `AUTO` does not currently inspect the surrounding input and dynamically
  decide to canonicalize such a matrix;
- `TRUE` explicitly requests canonicalization, while `FALSE` explicitly
  requests preservation of the input orientation.

Thus, for a general noncanonical matrix, `AUTO` and `FALSE` currently produce
the same geometry, but express different user intent. With #5648, `AUTO` would
fail early for `CELL_OPT`, whereas explicit `FALSE` would opt into the
unconstrained legacy behavior. Constrained cell optimizations would still be
rejected in that mode.

**Comment by DCM-Uni-Paderborn** [comment]
PR #5648 has now been merged as [`011c7422`](https://github.com/cp2k/cp2k/commit/011c742228f3c2b711d369ea9d44ae8f31c92370).

The final implementation supersedes my earlier description of #5648 as only a safety fix. `CELL_OPT` now optimizes the six independent deformation variables in a canonical frame and reconstructs the physical cell through a fixed rotation determined from the initial cell. This preserves the supplied cell orientation, atomic coordinates, trajectories, and restarts. Stress and anisotropic external pressure are transformed consistently into the optimization frame.

For `KEEP_SPACE_GROUP`, every proposed metric is projected onto the invariant metric of the detected lattice rotations while retaining its volume. The oriented primitive-silicon regression retains its input orientation and space group 227, and equivalent canonical and oriented inputs produce the same canonical optimization cell to numerical precision.

The merged change therefore addresses the original loss of space-group symmetry for oriented primitive cells. I am closing this discussion as resolved; the broader default-policy questions remain tracked in #5310.
